lib/libc/Kconfig

```diff
@@ -19,7 +19,8 @@ config PICOLIBC_SUPPORTED
 	bool
 	depends on ARC || ARM || ARM64 || MIPS || RISCV
 	depends on "$(ZEPHYR_TOOLCHAIN_VARIANT)" != "arcmwdt"
-	depends on !CPLUSPLUS || !PICOLIBC_MODULE
+        # Picolibc with C++ support in Zephyr SDK is handled by Zephyr SDK's own Kconfig.
+        depends on !(CPLUSPLUS && "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr")
```

Comment:
This definitely feels wrong -- `PICOLIBC_SUPPORTED` is true for everything except building with C++ and using picolibc from the module source. Every toolchain other than zephyr 0.15 and earlier may include picolibc, and may (or may not) include C++ support. I think we want to make the user responsible for determining whether the toolchain has the necessary bits in that case?

Reply:
Thanks for your feedback @keith-packard, highly appreciated so that we can get things right.

Especially because Kconfig and good dependencies can be tricky.

I tried to give the reason here:
https://github.com/zephyrproject-rtos/zephyr/pull/53338#pullrequestreview-1247175487

But let's take one step back, so that we are clear.

1. Picolibc, as built-in or as module, shall only be selectable if a toolchain supports picolibc.
   Support in this case can be supporting picolibc built as a module or if the toolchain has built-in support.
   1. arcmwdt does not support picolibc. Not built-in, nor as a as a module.
       (My understanding, please correct me if arcmwdt does support using picolibc as a module)
    1. Zephyr SDK 0.15 supports picolibc for C development
    2. Zephyr SDK >=0.16 supports picolibc for C and C++ development.
    3. Other 3rd party toolchains (crosstool-ng, arm embedded, etc) may support picolibc for C and C++ development.
        This depends on actual toolchain build configuration and / or version.
        Zephyr don't have toolchain testing in place for this kind of testing, so it's users responsibility to ensure picolibc is supported when using such toolchain. Thus Kconfig should always provide the possibility to select picolibc when such a toolchain is used.
2. If picolibc is supported, then user can decide if picolibc should be build as a module, or using the built-in version from the toolchain.
   For 3rd party toolchains, it's the users responsibility to know if toolchain supports built-in version, as described in 1.iv.
   

This gives us the `PICOLIBC` choice entry in Kconfig under `LIBC_IMPLEMENTATION` choice menu.

To minimize number of tests and dependencies in `PICOLIBC` entry we create the helper symbol `PICOLIBC_SUPPORTED`
This allows a simple:
```
config PICOLIBC
	bool "Picolibc library"
	depends on PICOLIBC_SUPPORTED
```

The `PICOLIBC_SUPPORTED` then defines if `PICOLIBC` should be available for user control.
Combining above, then this gives `PICOLIBC_SUPPORTED=y` in following cases:
`(3rd party toolchain && !arcmwdt) || (zephyr-sdk==0.15 && !C++) || (zephyr-sdk>=0.16)`

As a toolchain which is not arcmwdt and not Zephyr must be a 3rd party toolchain, then the above can be reduced to:
`(!arcmwdt && !zephyr-sdk) || (zephyr-sdk && !C++) 

Because ZEPHYR_TOOLCHAIN_VARIANT cannot be arcmwdt and Zephyr at the same time, the above can be written as: party toolchain, so first part of above can be written as:
`(!arcmwdt) && !(zephyr-sdk && C++)`

The last part:
`|| (zephyr-sdk>=0.16)` 
is what is included when Zephyr SDK 0.16 is included, cause then the Kconfig from there is sourced, and you'll have the following in Kconfig:
```
config PICOLIBC_SUPPORTED # From lib/libc/Kconfig
	bool
	depends on "$(ZEPHYR_TOOLCHAIN_VARIANT)" != "arcmwdt"
    depends on !(CPLUSPLUS && "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr")
	
config PICOLIBC_SUPPORTED # From zephyr-sdk-0.16/cmake/zephyr/Kconfig
    depends on "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr")
```

Two `depends on` in same setting translates into `&&`.
`depends on` in two configs of identical name translates into `||`.

An Zephyr-sdk 0.15 does not include the `zephyr-sdk-0.16/cmake/zephyr/Kconfig`, so therefore only the first part remains.

All of this only covers picolibc itself.

Whether to use picolibc as a module is then dependent on picolibc being used.
For example a user selecting `NEWLIB` should not be able to select `PICOLIBC_USE_MODULE`, and picolibc as a module should not be selectable if C++ is enabled (AFAIU).
And ofcourse it also depends on the Zephyr picolibc module being present in the workspace (west manifest)

Bullet 2. above therfore gives:
```
if PICOLIBC

config PICOLIBC_USE_MODULE
	bool "Picolibc as module"
	depends on ZEPHYR_PICOLIBC_MODULE
	depends on !CPLUSPLUS
```

No dependency loops in this description.
Of course both PICOLIBC when using a certain Zephyr SDK version depends on !C++ and so does the use of `PICOLIBC_USE_MODULE`, but that is not a loop.
A dependency loop in Kconfig forms if one of them suddenly starts to `select CPLUSPLUS`.

Of course if any assumptions in 1. and 2. are wrong, then the implementation is wrong.